console-subscriber/src/stats.rs

```diff
@@ -466,18 +462,49 @@ impl ToProto for ResourceStats {
 // === impl PollStats ===
 
 impl<H: RecordPoll> PollStats<H> {
-    fn start_poll(&self, at: Instant) {
-        if self.current_polls.fetch_add(1, AcqRel) == 0 {
-            // We are starting the first poll
-            let mut timestamps = self.timestamps.lock();
-            if timestamps.first_poll.is_none() {
-                timestamps.first_poll = Some(at);
-            }
+    fn wake(&self, at: Instant) {
+        let mut timestamps = self.timestamps.lock();
+        timestamps.last_wake = cmp::max(timestamps.last_wake, Some(at));
+    }
 
-            timestamps.last_poll_started = Some(at);
+    fn start_poll(&self, at: Instant) {
+        if self.current_polls.fetch_add(1, AcqRel) > 0 {
+            return;
+        }
 
-            self.polls.fetch_add(1, Release);
+        // We are starting the first poll
+        let mut timestamps = self.timestamps.lock();
+        if timestamps.first_poll.is_none() {
+            timestamps.first_poll = Some(at);
         }
+
+        timestamps.last_poll_started = Some(at);
+
+        self.polls.fetch_add(1, Release);
+
+        let scheduled = match (timestamps.last_wake, timestamps.last_poll_ended) {
+            // If the last poll ended after the last wake then it was likely
+            // a self-wake, so we measure from the end of the last poll instead.
+            // This also ensures that `busy_time` and `scheduled_time` don't overlap.
+            (Some(last_wake), Some(last_poll_ended)) if last_poll_ended > last_wake => {
+                last_poll_ended
+            }
+            (Some(last_wake), _) => last_wake,
+            (None, _) => return, // Async operations record polls, but not wakes
+        };
```

Comment:
nit, take it or leave it: I believe this is equivalent to just using the `cmp::PartialOrd` implementation for `Option`:
```suggestion
        // If the last poll ended after the last wake then it was likely
        // a self-wake, so we measure from the end of the last poll instead.
        // This also ensures that `busy_time` and `scheduled_time` don't overlap.
        let scheduled = match std::cmp::max(timestamps.last_wake, timestamps.last_poll_ended) {
            Some(scheduled) => scheduled,
            None => return, // Async operations record polls, but not wakes
        }
```